packages/taiko-client/prover/proof_producer/combined_producer.go

```diff
@@ -17,24 +17,143 @@ import (
 	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/metadata"
 )
 
+type BlockProofState struct {
+	verifiedTiers []uint16
+	proofs        []encoding.SubProof
+}
+
+type ProofStateManager struct {
+	mu     sync.Mutex
+	states map[uint64]*BlockProofState
+}
+
+func NewProofStateManager() *ProofStateManager {
+	return &ProofStateManager{
+		states: make(map[uint64]*BlockProofState),
+	}
+}
+
+func (m *ProofStateManager) create(blockID *big.Int) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		state = &BlockProofState{
+			verifiedTiers: []uint16{},
+			proofs:        []encoding.SubProof{},
+		}
+		m.states[blockIDUint64] = state
+	}
+}
+
+func (m *ProofStateManager) containsTier(blockID *big.Int, tier uint16) bool {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		return false
+	}
+
+	return slices.Contains(state.verifiedTiers, tier)
+}
+
+// addTierAndProof marks the tier as verified and adds the subproof to the block's state if
+// the state has not yet collected enough proofs. It returns whether the required number
+// of proofs has now been reached.
+func (m *ProofStateManager) addTierAndProof(
+	blockID *big.Int,
+	tier uint16,
+	subProof encoding.SubProof,
+	requiredProofs uint8,
+) (reachedRequired bool) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		// create if it doesn't exist
+		state = &BlockProofState{
+			verifiedTiers: []uint16{},
+			proofs:        []encoding.SubProof{},
+		}
+		m.states[blockIDUint64] = state
+	}
+
+	// Record that we've verified this tier
+	state.verifiedTiers = append(state.verifiedTiers, tier)
+
+	// Only add to the subproofs if we haven't reached requiredProofs count yet
+	if uint8(len(state.proofs)) < requiredProofs {
```

Comment:
Example for my comment above, you could tho this in go
```Go
if proofLimitNotReached := uint8(len(state.proofs)) < requiredProofs; proofLimitNotReached {}
```

This is what I mean give the context through the code. Although this is not the best example given how obvious if statement is, so I wouldn't bother for this particular case